File: basic_dsa_eighteen.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <math.h>

/**
 * Basic DSA Problems - Two More Fundamental Questions
 * Simple and essential DSA problems for beginners
 */

// Problem 1: Check if a Number is Strong Number
// A strong number is a number whose sum of factorial of digits equals the number itself
// Example: 145 = 1! + 4! + 5! = 1 + 24 + 120 = 145

/**
 * Calculate factorial of a number
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
long long factorial(int n) {
    if (n <= 1) {
        return 1;
    }
    long long result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

/**
 * Check if a number is strong number
 * Time Complexity: O(d * k) where d is number of digits, k is average digit value
 * Space Complexity: O(1)
 */
bool isStrongNumber(int n) {
    if (n <= 0) {
        return false;
    }
    
    int original = n;
    long long sumFactorial = 0;
    
    while (n > 0) {
        int digit = n % 10;
        sumFactorial += factorial(digit);
        n /= 10;
    }
    
    return sumFactorial == original;
}

/**
 * Get sum of factorials of each digit
 * Time Complexity: O(d * k) where d is number of digits
 * Space Complexity: O(1)
 */
long long getSumOfDigitFactorials(int n) {
    if (n <= 0) {
        return 0;
    }
    
    long long sum = 0;
    int temp = n;
    
    while (temp > 0) {
        int digit = temp % 10;
        sum += factorial(digit);
        temp /= 10;
    }
    
    return sum;
}

// Problem 2: Convert Binary to Decimal
// Convert a binary number to its decimal representation

/**
 * Convert binary string to decimal using iteration
 * Time Complexity: O(n) where n is length of binary string
 * Space Complexity: O(1)
 */
int binaryToDecimalIterative(const char* binaryStr) {
    if (binaryStr == NULL || strlen(binaryStr) == 0 || strcmp(binaryStr, "0") == 0) {
        return 0;
    }
    
    bool isNegative = binaryStr[0] == '-';
    int start = isNegative ? 1 : 0;
    int len = strlen(binaryStr);
    
    int decimal = 0;
    int power = 0;
    
    // Process from right to left (least significant bit to most)
    for (int i = len - 1; i >= start; i--) {
        if (binaryStr[i] == '1') {
            decimal += (int) pow(2, power);
        } else if (binaryStr[i] != '0') {
            printf("Error: Invalid binary digit: %c\n", binaryStr[i]);
            return -1;
        }
        power++;
    }
    
    return isNegative ? -decimal : decimal;
}

/**
 * Helper function for recursive binary conversion
 */
int binaryToDecimalRecursiveHelper(const char* binary, int index, int power) {
    if (index < 0) {
        return 0;
    }
    if (binary[index] == '1') {
        return (int) pow(2, power) + binaryToDecimalRecursiveHelper(binary, index - 1, power + 1);
    } else if (binary[index] == '0') {
        return binaryToDecimalRecursiveHelper(binary, index - 1, power + 1);
    } else {
        printf("Error: Invalid binary digit: %c\n", binary[index]);
        return -1;
    }
}

/**
 * Convert binary string to decimal using recursion
 * Time Complexity: O(n) where n is length of binary string
 * Space Complexity: O(n) due to recursion stack
 */
int binaryToDecimalRecursive(const char* binaryStr) {
    if (binaryStr == NULL || strlen(binaryStr) == 0 || strcmp(binaryStr, "0") == 0) {
        return 0;
    }
    
    bool isNegative = binaryStr[0] == '-';
    int start = isNegative ? 1 : 0;
    int len = strlen(binaryStr);
    
    int result = binaryToDecimalRecursiveHelper(binaryStr + start, len - start - 1, 0);
    return isNegative ? -result : result;
}

/**
 * Convert binary array to decimal
 * Time Complexity: O(n) where n is length of array
 * Space Complexity: O(1)
 */
int binaryArrayToDecimal(int* binaryArray, int length) {
    if (binaryArray == NULL || length == 0) {
        return 0;
    }
    
    int decimal = 0;
    int power = length - 1;
    
    for (int i = 0; i < length; i++) {
        if (binaryArray[i] != 0 && binaryArray[i] != 1) {
            printf("Error: Invalid binary digit: %d\n", binaryArray[i]);
            return -1;
        }
        decimal += binaryArray[i] * (int) pow(2, power);
        power--;
    }
    
    return decimal;
}

// Test cases
int main() {
    printf("============================================================\n");
    printf("Problem 1: Check if a Number is Strong Number\n");
    printf("============================================================\n");
    
    // Test Case 1 - Strong number 145
    int num1 = 145;
    bool result1 = isStrongNumber(num1);
    long long sum1 = getSumOfDigitFactorials(num1);
    printf("Number: %d\n", num1);
    printf("Sum of digit factorials: %lld\n", sum1);
    printf("Is strong number: %s\n\n", result1 ? "true" : "false");
    
    // Test Case 2 - Not strong number
    int num2 = 123;
    bool result2 = isStrongNumber(num2);
    long long sum2 = getSumOfDigitFactorials(num2);
    printf("Number: %d\n", num2);
    printf("Sum of digit factorials: %lld\n", sum2);
    printf("Is strong number: %s\n\n", result2 ? "true" : "false");
    
    // Test Case 3 - Strong number 1
    int num3 = 1;
    bool result3 = isStrongNumber(num3);
    printf("Number: %d\n", num3);
    printf("Is strong number: %s\n\n", result3 ? "true" : "false");
    
    // Test Case 4 - Strong number 2
    int num4 = 2;
    bool result4 = isStrongNumber(num4);
    printf("Number: %d\n", num4);
    printf("Is strong number: %s\n\n", result4 ? "true" : "false");
    
    // Test Case 5 - Not strong number
    int num5 = 10;
    bool result5 = isStrongNumber(num5);
    long long sum5 = getSumOfDigitFactorials(num5);
    printf("Number: %d\n", num5);
    printf("Sum of digit factorials: %lld\n", sum5);
    printf("Is strong number: %s\n\n", result5 ? "true" : "false");
    
    printf("============================================================\n");
    printf("Problem 2: Convert Binary to Decimal\n");
    printf("============================================================\n");
    
    // Test Case 1 - Basic binary
    const char* binary1 = "1010";
    int result1_dec = binaryToDecimalIterative(binary1);
    int result1_rec = binaryToDecimalRecursive(binary1);
    printf("Binary: %s\n", binary1);
    printf("Decimal (iterative): %d\n", result1_dec);
    printf("Decimal (recursive): %d\n\n", result1_rec);
    
    // Test Case 2 - Zero
    const char* binary2 = "0";
    int result2_dec = binaryToDecimalIterative(binary2);
    printf("Binary: %s\n", binary2);
    printf("Decimal: %d\n\n", result2_dec);
    
    // Test Case 3 - Large binary
    const char* binary3 = "11111111";
    int result3_dec = binaryToDecimalIterative(binary3);
    printf("Binary: %s\n", binary3);
    printf("Decimal: %d\n\n", result3_dec);
    
    // Test Case 4 - Binary array
    int binary4[] = {1, 0, 1, 1, 0, 1};
    int length4 = sizeof(binary4) / sizeof(binary4[0]);
    int result4_dec = binaryArrayToDecimal(binary4, length4);
    printf("Binary (array): ");
    for (int i = 0; i < length4; i++) {
        printf("%d", binary4[i]);
    }
    printf("\nDecimal: %d\n\n", result4_dec);
    
    // Test Case 5 - Single bit
    const char* binary5 = "1";
    int result5_dec = binaryToDecimalIterative(binary5);
    printf("Binary: %s\n", binary5);
    printf("Decimal: %d\n\n", result5_dec);
    
    return 0;
}

File: basic_dsa_eighteen.py
"""
Basic DSA Problems - Two More Fundamental Questions
Simple and essential DSA problems for beginners
"""


# Problem 1: Check if a Number is Strong Number
# A strong number is a number whose sum of factorial of digits equals the number itself
# Example: 145 = 1! + 4! + 5! = 1 + 24 + 120 = 145

def factorial(n):
    """
    Calculate factorial of a number
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if n <= 1:
        return 1
    result = 1
    for i in range(2, n + 1):
        result *= i
    return result


def is_strong_number(n):
    """
    Check if a number is strong number
    Time Complexity: O(d * k) where d is number of digits, k is average digit value
    Space Complexity: O(1)
    """
    if n <= 0:
        return False
    
    original = n
    sum_factorial = 0
    
    while n > 0:
        digit = n % 10
        sum_factorial += factorial(digit)
        n //= 10
    
    return sum_factorial == original


def get_digit_factorials(n):
    """
    Get list of factorials of each digit
    Time Complexity: O(d * k) where d is number of digits
    Space Complexity: O(d)
    """
    if n <= 0:
        return []
    
    factorials = []
    temp = n
    
    while temp > 0:
        digit = temp % 10
        factorials.insert(0, factorial(digit))
        temp //= 10
    
    return factorials


# Problem 2: Convert Binary to Decimal
# Convert a binary number to its decimal representation

def binary_to_decimal_iterative(binary_str):
    """
    Convert binary string to decimal using iteration
    Time Complexity: O(n) where n is length of binary string
    Space Complexity: O(1)
    """
    if not binary_str or binary_str == "0":
        return 0
    
    is_negative = binary_str[0] == '-'
    start = 1 if is_negative else 0
    
    decimal = 0
    power = 0
    
    # Process from right to left (least significant bit to most)
    for i in range(len(binary_str) - 1, start - 1, -1):
        if binary_str[i] == '1':
            decimal += 2 ** power
        elif binary_str[i] != '0':
            raise ValueError(f"Invalid binary digit: {binary_str[i]}")
        power += 1
    
    return -decimal if is_negative else decimal


def binary_to_decimal_recursive(binary_str):
    """
    Convert binary string to decimal using recursion
    Time Complexity: O(n) where n is length of binary string
    Space Complexity: O(n) due to recursion stack
    """
    def helper(binary, index, power):
        if index < 0 or (index == 0 and binary[0] == '-'):
            return 0
        if binary[index] == '1':
            return (2 ** power) + helper(binary, index - 1, power + 1)
        elif binary[index] == '0':
            return helper(binary, index - 1, power + 1)
        else:
            raise ValueError(f"Invalid binary digit: {binary[index]}")
    
    if not binary_str or binary_str == "0":
        return 0
    
    is_negative = binary_str[0] == '-'
    start = len(binary_str) - 1
    
    result = helper(binary_str, start, 0)
    return -result if is_negative else result


def binary_to_decimal_builtin(binary_str):
    """
    Convert binary string to decimal using built-in function
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if not binary_str:
        return 0
    
    is_negative = binary_str[0] == '-'
    binary = binary_str[1:] if is_negative else binary_str
    
    decimal = int(binary, 2)
    return -decimal if is_negative else decimal


def binary_array_to_decimal(binary_array):
    """
    Convert binary array to decimal
    Time Complexity: O(n) where n is length of array
    Space Complexity: O(1)
    """
    if not binary_array:
        return 0
    
    decimal = 0
    power = len(binary_array) - 1
    
    for digit in binary_array:
        if digit != 0 and digit != 1:
            raise ValueError(f"Invalid binary digit: {digit}")
        decimal += digit * (2 ** power)
        power -= 1
    
    return decimal


# Test cases
if __name__ == "__main__":
    print("=" * 60)
    print("Problem 1: Check if a Number is Strong Number")
    print("=" * 60)
    
    # Test Case 1 - Strong number 145
    num1 = 145
    result1 = is_strong_number(num1)
    factorials1 = get_digit_factorials(num1)
    print(f"Number: {num1}")
    print(f"Digit factorials: {factorials1}")
    print(f"Sum of factorials: {sum(factorials1)}")
    print(f"Is strong number: {result1}")
    
    # Test Case 2 - Not strong number
    num2 = 123
    result2 = is_strong_number(num2)
    factorials2 = get_digit_factorials(num2)
    print(f"\nNumber: {num2}")
    print(f"Digit factorials: {factorials2}")
    print(f"Sum of factorials: {sum(factorials2)}")
    print(f"Is strong number: {result2}")
    
    # Test Case 3 - Strong number 1
    num3 = 1
    result3 = is_strong_number(num3)
    print(f"\nNumber: {num3}")
    print(f"Is strong number: {result3}")
    
    # Test Case 4 - Strong number 2
    num4 = 2
    result4 = is_strong_number(num4)
    print(f"\nNumber: {num4}")
    print(f"Is strong number: {result4}")
    
    # Test Case 5 - Not strong number
    num5 = 10
    result5 = is_strong_number(num5)
    factorials5 = get_digit_factorials(num5)
    print(f"\nNumber: {num5}")
    print(f"Digit factorials: {factorials5}")
    print(f"Sum of factorials: {sum(factorials5)}")
    print(f"Is strong number: {result5}")
    
    print("\n" + "=" * 60)
    print("Problem 2: Convert Binary to Decimal")
    print("=" * 60)
    
    # Test Case 1 - Basic binary
    binary1 = "1010"
    result1_iter = binary_to_decimal_iterative(binary1)
    result1_rec = binary_to_decimal_recursive(binary1)
    result1_builtin = binary_to_decimal_builtin(binary1)
    print(f"Binary: {binary1}")
    print(f"Decimal (iterative): {result1_iter}")
    print(f"Decimal (recursive): {result1_rec}")
    print(f"Decimal (built-in): {result1_builtin}")
    
    # Test Case 2 - Zero
    binary2 = "0"
    result2 = binary_to_decimal_iterative(binary2)
    print(f"\nBinary: {binary2}")
    print(f"Decimal: {result2}")
    
    # Test Case 3 - Large binary
    binary3 = "11111111"
    result3 = binary_to_decimal_iterative(binary3)
    print(f"\nBinary: {binary3}")
    print(f"Decimal: {result3}")
    
    # Test Case 4 - Binary array
    binary4 = [1, 0, 1, 1, 0, 1]
    result4 = binary_array_to_decimal(binary4)
    print(f"\nBinary (array): {binary4}")
    print(f"Decimal: {result4}")
    
    # Test Case 5 - Single bit
    binary5 = "1"
    result5 = binary_to_decimal_iterative(binary5)
    print(f"\nBinary: {binary5}")
    print(f"Decimal: {result5}")
